src/import/load-costume.js

```diff
@@ -30,76 +30,151 @@ const loadVector_ = function (costume, costumeAsset, runtime, rotationCenter, op
+            }
+            canvas = runtime.v2BitmapAdapter.resize(canvas, canvas.width * scale, canvas.height * scale);
         }
-    }).then(imageElement => {
-        // createBitmapSkin does the right thing if costume.bitmapResolution or rotationCenter are undefined...
-        costume.skinId = runtime.renderer.createBitmapSkin(imageElement, costume.bitmapResolution, rotationCenter);
-        const renderSize = runtime.renderer.getSkinSize(costume.skinId);
-        costume.size = [renderSize[0] * 2, renderSize[1] * 2]; // Actual size, since all bitmaps are resolution 2
-
-        if (!rotationCenter) {
-            rotationCenter = runtime.renderer.getSkinRotationCenter(costume.skinId);
-            // Actual rotation center, since all bitmaps are resolution 2
-            costume.rotationCenterX = rotationCenter[0] * 2;
-            costume.rotationCenterY = rotationCenter[1] * 2;
-            costume.bitmapResolution = 2;
+
+        // By scaling, we've converted it to bitmap resolution 2
+        if (rotationCenter) {
+            rotationCenter[0] = rotationCenter[0] * scale;
+            rotationCenter[1] = rotationCenter[1] * scale;
+            costume.rotationCenterX = rotationCenter[0];
+            costume.rotationCenterY = rotationCenter[1];
         }
-        return costume;
-    });
+        costume.bitmapResolution = 2;
+
+        return {
+            canvas: canvas,
+            rotationCenter: rotationCenter,
+            // True if the asset matches the base layer; false if it required adjustment
+            assetMatchesBase: scale === 1 && !textImageElement
+        };
+    })
+        .catch(e => Promise.reject(e))
+        .finally(() => {
+            // Clean up the costume object
+            delete costume.textLayerMD5;
+            delete costume.textLayerAsset;
+        });
+};
+
+const loadBitmap_ = function (costume, runtime, rotationCenter) {
+    return fetchBitmapCanvas_(costume, runtime, rotationCenter).then(fetched => new Promise(resolve => {
+        rotationCenter = fetched.rotationCenter;
+
+        const updateCostumeAsset = function (dataURI) {
+            if (!runtime.v2BitmapAdapter) {
+                return Promise.reject('No V2 Bitmap adapter present.');
+            }
+
+            const storage = runtime.storage;
+            costume.asset = storage.createAsset(
+                storage.AssetType.ImageBitmap,
+                storage.DataFormat.PNG,
+                runtime.v2BitmapAdapter.convertDataURIToBinary(dataURI),
+                null,
+                true // generate md5
+            );
+            costume.assetId = costume.asset.assetId;
+            costume.md5 = `${costume.assetId}.${costume.dataFormat}`;
+        };
+
+        if (!fetched.assetMatchesBase) {
+            updateCostumeAsset(fetched.canvas.toDataURL());
+        }
+        resolve(fetched.canvas);
+    }))
+        .catch(e => Promise.reject(e))
```

Comment:
Oh wow, I didn't know you could put a catch in the middle of a then chain... does it make a difference being in the middle rather than at the end?

Reply:
The catch documentation says,

> It behaves the same as calling Promise.prototype.then(undefined, onRejected) (in fact, calling obj.catch(onRejected) internally calls obj.then(undefined, onRejected)).

This is what I was asking about in comment https://github.com/LLK/scratch-vm/pull/1737#issuecomment-438817340 , I think I don't actually need the catch here, but I don't understand how promise rejection works. Say the promise right before the catch here rejects. Then does the rejection bubble all the way up the chain automatically? Or do I need to manually pass the rejection back to the previous then, using this catch, otherwise it will return Promise.resolve(null)?

Reply:
The rejection should bubble up the chain automatically. Is there a test we can set up that ensures this?

Reply:
I want to add a test for loadCostume but I can't in VM because tests in VM don't recognize browser elements. Hopefully I can add one in GUI.

I made a little JS fiddle that shows that the error bubbles up through thens, even if one of the thens in the chain explicitly catches the error and chooses to ignore it... so I'll remove these catches that do nothing.
https://jsfiddle.net/c298frvb/